collections/src/main/scala/strawman/collection/immutable/LazyList.scala

```diff
@@ -225,10 +225,24 @@ sealed abstract class LazyList[+A] extends LinearSeq[A] with LazyListOps[A, Lazy
     if (this.isEmpty) z
     else tail.foldLeft(op(z, head))(op)
   }
+
+  override def toString: String = {
+    @tailrec
+    def _toString(ll: LazyList[A], acc: List[String]): List[String] = ll.memberEvaluationState match {
+      case None                 => acc
+      case Some((true, false))  => "?" :: s"${ll.head}" :: acc
+      case Some((true, true))   => _toString(ll.tail, s"${ll.head}" :: acc)
+      case Some((false, false)) => "?" :: acc
+      case Some((false, true))  => _toString(ll.tail, "?" :: acc)
+    }
+
+    s"$className${_toString(this, List()).reverse.mkString("(", ", ", ")")}"
```

Comment:
I think you can find an algorithm that does not require building a list, reversing it and then computing the string from it. As @marcelocenerine suggested, you can draw inspiration from the current `Iterator.addString` implementation, and the old `Stream.toString` implementation.

Reply:
Here, I was just playing around with an idea of tail-recursive solution using `List` as result accumulator in this commit. In recent commits, I have made changes by using `mutable.StringBuilder` instead of `List`. So it should be in good in terms of performance over constructing a list and then reversing it.